clients/src/main/java/org/apache/kafka/common/metrics/stats/TokenBucket.java

```diff
@@ -0,0 +1,127 @@
+ * - T3 => Total at T+6s => max(0, T2 - Q * dT) = (6 - 2 * 2) = 2
+ */
+public class TokenBucket extends SampledStat {
+
+    private final TimeUnit unit;
+
+    public TokenBucket() {
+        this(TimeUnit.SECONDS);
+    }
+
+    public TokenBucket(TimeUnit unit) {
+        super(0);
+        this.unit = unit;
+    }
+
+    @Override
+    protected void update(Sample sample, MetricConfig config, double value, long now) {
+        sample.value += value;
+    }
+
+    @Override
+    public double combine(List<Sample> samples, MetricConfig config, long now) {
+        if (this.samples.isEmpty())
+            return 0;
+
+        final double quota = config.quota() != null ? config.quota().bound() : 0;
+        final int startIndex = (this.current + 1) % this.samples.size();
+
+        long lastWindowMs = Long.MAX_VALUE;
+        double total = 0.0;
+
+        for (int i = 0; i < this.samples.size(); i++) {
+            int current = (startIndex + i) % this.samples.size();
+            Sample sample = this.samples.get(current);
+
+            // The total is decreased by (last window - current window) * quota. We basically
+            // pay back the amount of credits spent until we reach zero (dept fully paid).
+            total = adjustedTotal(total, quota, lastWindowMs, sample.lastWindowMs);
+
+            // The current sample is added to the total. If the sample is a negative number,
+            // we ensure that total remains >= 0. A negative number can be used to give back
+            // credits if there were not used. If the number of credits given back is higher
+            // than the current dept, we basically assume that the dept is paid off.
+            total = Math.max(0, total + sample.value);
+
+            // Memorize the current window so we can use for the next sample.
+            lastWindowMs = sample.lastWindowMs;
+        }
+
+        return adjustedTotal(total, quota, lastWindowMs, now);
+    }
+
+    private double adjustedTotal(double total, double quota, long lastWindowMs, long nowMs) {
+        return Math.max(0, total - delta(quota, lastWindowMs, nowMs));
+    }
+
+    private double delta(double quota, long lastWindowMs, long nowMs) {
+        return quota * convert(Math.max(0, nowMs - lastWindowMs));
+    }
+
+    private double convert(long timeMs) {
+        switch (unit) {
+            case NANOSECONDS:
+                return timeMs * 1000.0 * 1000.0;
+            case MICROSECONDS:
+                return timeMs * 1000.0;
+            case MILLISECONDS:
+                return timeMs;
+            case SECONDS:
+                return timeMs / 1000.0;
+            case MINUTES:
+                return timeMs / (60.0 * 1000.0);
+            case HOURS:
+                return timeMs / (60.0 * 60.0 * 1000.0);
+            case DAYS:
+                return timeMs / (24.0 * 60.0 * 60.0 * 1000.0);
+            default:
+                throw new IllegalStateException("Unknown unit: " + unit);
+        }
```

Comment:
why not just use `TimeUnit.convert()`?

Reply:
I just copy-pasted the code from another internal class and I have to admit that I was not aware of `TimeUnit.convert()`. Thanks for pointing that out.